feat: define access/claim in @web3-storage/capabilities

gobengo opened the pull request:
Motivation:
* https://github.com/web3-storage/w3protocol/issues/408

Unblocks:
* people importing a canonical CapabilityParser for `access/claim`
* adding advanced tests e.g. by [reverting this and finishing the TODOs](https://github.com/web3-storage/w3protocol/pull/409/commits/fd0b56ff7459d0178e3b9daa64a59bf1bcd9a463)
  * these involve `can=./update` abilities, so may make most sense to add after https://github.com/web3-storage/w3protocol/pull/392 which defines tools for them

gobengo commented on packages/capabilities/src/access.js:
```diff
@@ -100,3 +100,12 @@ export const session = capability({
+    with: DID.match({ method: 'key' }).or(DID.match({ method: 'mailto' })),
+    derives: equalWith,
+  }),
+  derives: equalWith,
```
could use feedback ensuring these two `derives: equalWith` make sense

Gozala replied:
The first one is obsolete, because omitting `derives` there would basically use `equalWith`. Second one is mandatory and is used to compare `claim` with `base`. Given that `claim` has no `nb` comparing just `with` fields seems like a right to me.

Gozala reviewed (approved):
Changes look good, but I think it would be better to add more tests and an `access/delegate` capability before landing.

That said I'm fine landing this as is and doing rest as followups.

gobengo commented on packages/capabilities/src/access.js:
```diff
@@ -100,3 +100,12 @@ export const session = capability({
   },
 })
+
+export const claim = base.derive({
```
test:
* delegation works
  * from access/*
  * or from *
* can self-issue cap

gobengo commented on packages/capabilities/test/capabilities/access.test.js:
```diff
@@ -222,4 +223,26 @@ describe('access capabilities', function () {
+        await example({ issuer: bob })
+      }
+    })
+  })
```
Would you say this is a valid test for what you mean by #1?
* https://github.com/web3-storage/w3protocol/pull/409/commits/1a0016c825c1fceeea4f98c81dd8266b31f104b5#diff-67e9cd2926c4b07ca092fa34e8015b5ed3ce81c5708473253964f5d7a0da0c4fR313

For #2 I added:
* throws if try to create invocation using `Access.claim` using unexpected `with` did method - https://github.com/web3-storage/w3protocol/pull/409/commits/6556e21fff69bd007dc3cbee77011855de41ebcc#diff-67e9cd2926c4b07ca092fa34e8015b5ed3ce81c5708473253964f5d7a0da0c4fR268
* does not authorize `access/claim` whose `with` is an unexpected did method https://github.com/web3-storage/w3protocol/pull/409/commits/a57ef04215deacb97ea41c842cdda3963d3a9a10#diff-67e9cd2926c4b07ca092fa34e8015b5ed3ce81c5708473253964f5d7a0da0c4fR282

gobengo commented:
update
* added some tests suggested by @Gozala https://github.com/web3-storage/w3protocol/pull/409#discussion_r1092585822
* added tests from discussing with @hugomrdias 
  * can derive access/claim https://github.com/web3-storage/w3protocol/pull/409/commits/aceffc24c0569284248bd80918ab7deca3272210
  * when testing use `import('@ucanto/validator').access` to make assertions about how the capabilityParsers will affect authorization https://github.com/web3-storage/w3protocol/pull/409/commits/e6b6963e7c0fab58cd6648e7ff4a71430f8cdaa8#diff-67e9cd2926c4b07ca092fa34e8015b5ed3ce81c5708473253964f5d7a0da0c4fR251

gobengo commented on packages/capabilities/test/capabilities/access.test.js:
```diff
@@ -3,6 +3,8 @@ import { access } from '@ucanto/validator'
 import { Verifier } from '@ucanto/principal/ed25519'
 import * as Access from '../../src/access.js'
 import { alice, bob, service, mallory } from '../helpers/fixtures.js'
+import * as Ucanto from '@ucanto/interface'
```
This would have required a comment to disable tsc no-unused-vars, but it no longer does because of [adding jsdoc/no-undefined-types rule](https://github.com/web3-storage/w3protocol/pull/409/files#diff-ba437578db4413909c1b02066b4d818ccb40dc79932f754051415b17f39a150aR92)

related to: https://github.com/web3-storage/w3protocol/pull/392/files#r1090596726
